enum.c

```diff
@@ -1385,6 +1385,205 @@ sort_by_cmp(const void *ap, const void *bp, void *data)
+    }
+    for (size_t offset = n - 1; offset > 0;)
+    {
+        VALUE tmp_1 = *ptr_begin;
+        VALUE tmp_2 = *(ptr_begin + 1);
+
+        *ptr_begin = *(ptr_begin + (offset << 1));
+        *(ptr_begin + 1) =  *(ptr_begin + (offset << 1) + 1);
+
+        *(ptr_begin + (offset << 1)) = tmp_1;
+        *(ptr_begin + (offset << 1) + 1) = tmp_2;
+
+        rb_uniform_heap_down_2(ptr_begin, 0, --offset);
+    }
+}
+
+
+static void
+rb_uniform_quicksort_intro_2(VALUE* ptr_begin, VALUE *ptr_end, VALUE d)
+{
+
+    if (ptr_end - ptr_begin <= 32)
+    {
+        rb_uniform_insertionsort_2(ptr_begin, ptr_end);
+        return;
+    }
+    if (d == 0)
+    {
+        rb_uniform_heapsort_2(ptr_begin, ptr_end);
+        return;
+    }
+
+    VALUE x;
+    size_t offset = ptr_end - ptr_begin;
+    offset = offset >> 1;
+    if (offset & 1)
+    {
+        offset += 1;
+    }
+    x = med3_val(*ptr_begin, *(ptr_begin + offset), *(ptr_end - 2));
+
+
+    VALUE *i = ptr_begin;
+    VALUE *j = ptr_end - 2;
+    do
+    {
+        while (ucmp(*i, x)) i += 2;
+        while (ucmp(x, *j)) j -= 2;
+        if (i <= j)
+        {
+
+            VALUE tmp_1 = *i;
+            VALUE tmp_2 = *(i + 1);
+
+            *i = *j;
+            *(i + 1) = *(j + 1);
+
+            *j = tmp_1;
+            *(j + 1) = tmp_2;
+
+            i += 2;
+            j -= 2;
+        }
+    }
+    while (i <= j);
+    j += 2;
+    if (ptr_end - j > 2)   rb_uniform_quicksort_intro_2(j, ptr_end, d - 1);
+    if (i - ptr_begin > 2) rb_uniform_quicksort_intro_2(ptr_begin, i, d - 1);
+}
+
+/**
+ * Direct primitive data compare sort. Implement with intro sort.
+ * @param[in]     ptr_begin  The begin address of target rb_ary's raw pointer.
+ * @param[in]     ptr_end    The end address of target rb_ary's raw pointer.
+**/
+void
+rb_uniform_intro_sort_2(VALUE * ptr_begin, VALUE *ptr_end)
+{
+    VALUE d = 0;
+    VALUE n = (ptr_end - ptr_begin) >> 1;
+    for (d = 0; n > 1; n >>= 1) d++;
+
+    bool sorted_flag = true;
+    for (VALUE *ptr = ptr_begin + 2; ptr < ptr_end; ptr += 2)
+    {
+        if ((OPTIMIZED_CMP(*(ptr - 2), *ptr)) == 1)
+        {
+            sorted_flag = false;
+            break;
+        }
+    }
+
+    if (sorted_flag)
+    {
+        return;
+    }
+    rb_uniform_quicksort_intro_2(ptr_begin, ptr_end, d << 1);
+}
+
+#undef ucmp
+
+/**
+ * Scans a rb_ary, judge if elements in rb_ary are uniformed to fixnum or float.
+ * @param[in]     ary        Target
+ * @return        True, if elements are uniformed to fixnum or float. False if
+                  either one of element is not fixnum or float.
+**/
+bool
+rb_uniform_only_num_2(VALUE ary)
+{
+
+    long len = RARRAY_LEN(ary);
+
+    const VALUE *ptr = RARRAY_CONST_PTR_TRANSIENT(ary);
+
+    for (long offset = 0; offset < len; offset += 2)
+    {
+        if ((!FIXNUM_P(ptr[offset])) && (!RB_FLOAT_TYPE_P(ptr[offset])))
+        {
+            return false;
+        }
+    }
+    return true;
+}
+
+
```

Comment:
If checking the result from the given block in `sort_by_i`, this function is not needed.
`struct sort_by_data::n` should not exceed `SORT_BY_BUFSIZE` (= 16), it needs only 5bits.

Reply:
You means use 1 bit in `struct sort_by_data::n` to make a flag for uniform?

Reply:
Oh, maybe not so much overhead

I implement like flow
 
in `sort_by_i `
```
if (data->n & SORT_BY_UNIFORM_MASK)
    {
         data->n = data->n & (~((!FIXNUM_P(v) && (!RB_FLOAT_TYPE_P(v))) << SORT_BY_UNIFORM_LSHIFT));
    }

    RARRAY_ASET(data->buf, (data->n & SORT_BY_BUFFSIZE_MASK)*2, v);
    RARRAY_ASET(data->buf, (data->n*2 & SORT_BY_BUFFSIZE_MASK)+1, i);
    data->n++;
    if ((data->n & SORT_BY_BUFFSIZE_MASK) == SORT_BY_BUFSIZE) {
        rb_ary_concat(ary, data->buf);
        data->n = data->n & SORT_BY_UNIFORM_MASK;
    }
```

`in enum_sort_by`

```
data->n = SORT_BY_UNIFORM_MASK;
    rb_block_call(obj, id_each, 0, 0, sort_by_i, (VALUE)memo);
    ary = data->ary;
    buf = data->buf;
    if (data->n & SORT_BY_BUFFSIZE_MASK) {
        rb_ary_resize(buf, (data->n & SORT_BY_BUFFSIZE_MASK)*2);
        rb_ary_concat(ary, buf);
    }
    if (RARRAY_LEN(ary) > 2) {
        if (data->n & SORT_BY_UNIFORM_MASK)
```

And run mini bench-mark

`ruby 3.1.2p20 (2022-04-12 revision 4491bb740a) [x64-mingw-ucrt]`
```
            user     system      total        real
Fixnum Array Sort  0.109000   0.000000   0.109000 (  0.110664)
Fixnum Array Sort by  0.250000   0.016000   0.266000 (  0.254851)
            user     system      total        real
Float Array Sort  0.204000   0.000000   0.204000 (  0.208649)
Float Array Sort by  0.343000   0.015000   0.358000 (  0.358526)
            user     system      total        real
String length Sort by  0.250000   0.000000   0.250000 (  0.263366)
            user     system      total        real
String Sort by  5.422000   0.000000   5.422000 (  5.421803)

```

`ruby 3.3.0dev (2023-05-13T03:40:36Z master d9a0d418cd) [x64-mingw32]` with bit operation
```
            user     system      total        real
Fixnum Array Sort  0.109000   0.000000   0.109000 (  0.104900)
Fixnum Array Sort by  0.094000   0.000000   0.094000 (  0.102939)
            user     system      total        real
Float Array Sort  0.187000   0.000000   0.187000 (  0.187115)
Float Array Sort by  0.203000   0.000000   0.203000 (  0.205943)
            user     system      total        real
String length Sort by  0.125000   0.000000   0.125000 (  0.116119)
            user     system      total        real
String length Sort by  5.454000   0.000000   5.454000 (  5.452588)

``

Reply:
If this kind of implement  is ok, I will commit it.